File: flutter_app/lib/services/api_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'dart:typed_data';
import 'package:flutter/foundation.dart';
import '../models/user.dart';
import '../models/vehicle.dart';

class ApiService {
  static const String baseUrl = 'https://vehicleq.onrender.com'; // Production backend URL

  // User Registration
  static Future<User> registerUser({
    required String username,
    required String email,
    required String password,
    required String fullName,
    required String phone,
  }) async {
    try {
      final request = http.MultipartRequest(
        'POST',
        Uri.parse('$baseUrl/register/'),
      );

      request.fields['username'] = username;
      request.fields['email'] = email;
      request.fields['password'] = password;
      request.fields['full_name'] = fullName;
      request.fields['phone'] = phone;

      final response = await request.send();
      final responseData = await response.stream.bytesToString();

      if (response.statusCode == 200) {
        return User.fromJson(jsonDecode(responseData));
      } else {
        throw Exception('Failed to register: ${jsonDecode(responseData)['detail']}');
      }
    } catch (e) {
      throw Exception('Registration error: $e');
    }
  }

  // User Login
  static Future<User> loginUser({
    required String username,
    required String password,
  }) async {
    try {
      final request = http.MultipartRequest(
        'POST',
        Uri.parse('$baseUrl/login/'),
      );

      request.fields['username'] = username;
      request.fields['password'] = password;

      final response = await request.send();
      final responseData = await response.stream.bytesToString();

      if (response.statusCode == 200) {
        return User.fromJson(jsonDecode(responseData));
      } else {
        throw Exception('Invalid username or password');
      }
    } catch (e) {
      throw Exception('Login error: $e');
    }
  }

  // Get User Profile
  static Future<User> getUserProfile(int userId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/profile/$userId'),
      );

      if (response.statusCode == 200) {
        return User.fromJson(jsonDecode(response.body));
      } else {
        throw Exception('Failed to fetch profile');
      }
    } catch (e) {
      throw Exception('Error fetching profile: $e');
    }
  }

  // Update User Profile
  static Future<User> updateProfile({
    required int userId,
    required String fullName,
    required String phone,
  }) async {
    try {
      final request = http.MultipartRequest(
        'PUT',
        Uri.parse('$baseUrl/profile/$userId'),
      );

      request.fields['full_name'] = fullName;
      request.fields['phone'] = phone;

      final response = await request.send();
      final responseData = await response.stream.bytesToString();

      if (response.statusCode == 200) {
        return User.fromJson(jsonDecode(responseData));
      } else {
        throw Exception('Failed to update profile');
      }
    } catch (e) {
      throw Exception('Error updating profile: $e');
    }
  }

  // Get All Vehicles
  static Future<List<Vehicle>> getVehicles() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/vehicles/'),
      );

      if (response.statusCode == 200) {
        List<dynamic> data = jsonDecode(response.body);
        return data.map((v) => Vehicle.fromJson(v)).toList();
      } else {
        throw Exception('Failed to fetch vehicles');
      }
    } catch (e) {
      throw Exception('Error fetching vehicles: $e');
    }
  }

  // Get User's Vehicles
  static Future<List<Vehicle>> getUserVehicles(int userId) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/vehicles/$userId'),
      );

      if (response.statusCode == 200) {
        List<dynamic> data = jsonDecode(response.body);
        return data.map((v) => Vehicle.fromJson(v)).toList();
      } else {
        throw Exception('Failed to fetch user vehicles');
      }
    } catch (e) {
      throw Exception('Error fetching vehicles: $e');
    }
  }

  // Upload Vehicle
  static Future<Vehicle> uploadVehicle({
    required int userId,
    required String vehicleNumber,
    required String owner,
    String? imagePath,
    Uint8List? imageBytes,
    String? filename,
  }) async {
    try {
      final request = http.MultipartRequest(
        'POST',
        Uri.parse('$baseUrl/upload/'),
      );

      request.fields['user_id'] = userId.toString();
      request.fields['number'] = vehicleNumber;
      request.fields['owner'] = owner;
      if (imageBytes != null) {
        final name = filename ?? (imagePath != null ? imagePath.split('/').last : 'upload.jpg');
        request.files.add(
          http.MultipartFile.fromBytes('image', imageBytes, filename: name),
        );
      } else if (imagePath != null) {
        request.files.add(
          await http.MultipartFile.fromPath('image', imagePath),
        );
      } else {
        throw Exception('No image provided for upload');
      }

      final response = await request.send();
      final responseData = await response.stream.bytesToString();

      if (response.statusCode == 200) {
        return Vehicle.fromJson(jsonDecode(responseData));
      } else {
        throw Exception('Failed to upload vehicle');
      }
    } catch (e) {
      throw Exception('Upload error: $e');
    }
  }

  // Delete Vehicle
  static Future<bool> deleteVehicle(int vehicleId) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/vehicle/$vehicleId'),
      );

      if (response.statusCode == 200) {
        return true;
      } else {
        throw Exception('Failed to delete vehicle');
      }
    } catch (e) {
      throw Exception('Error deleting vehicle: $e');
    }
  }

  // Get Image URL
  static String getImageUrl(int vehicleId) {
    return '$baseUrl/image/$vehicleId';
  }
}

File: angular_app/src/app/services/api.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { User } from '../models/user.model';
import { Vehicle } from '../models/vehicle.model';

@Injectable({
  providedIn: 'root'
})
export class ApiService {
  // Production URL
  private baseUrl = 'https://vehicleq.onrender.com';

  constructor(private http: HttpClient) { }

  // User Registration
  registerUser(username: string, email: string, password: string, fullName: string, phone: string): Observable<User> {
    const formData = new FormData();
    formData.append('username', username);
    formData.append('email', email);
    formData.append('password', password);
    formData.append('full_name', fullName);
    formData.append('phone', phone);

    return this.http.post<User>(`${this.baseUrl}/register/`, formData);
  }

  // User Login
  loginUser(username: string, password: string): Observable<User> {
    const formData = new FormData();
    formData.append('username', username);
    formData.append('password', password);

    return this.http.post<User>(`${this.baseUrl}/login/`, formData);
  }

  // Get User Profile
  getUserProfile(userId: number): Observable<User> {
    return this.http.get<User>(`${this.baseUrl}/profile/${userId}`);
  }

  // Update User Profile
  updateProfile(userId: number, fullName: string, phone: string): Observable<User> {
    const formData = new FormData();
    formData.append('full_name', fullName);
    formData.append('phone', phone);

    return this.http.put<User>(`${this.baseUrl}/profile/${userId}`, formData);
  }

  // Get All Vehicles
  getVehicles(): Observable<Vehicle[]> {
    return this.http.get<Vehicle[]>(`${this.baseUrl}/vehicles/`);
  }

  // Get User's Vehicles
  getUserVehicles(userId: number): Observable<Vehicle[]> {
    return this.http.get<Vehicle[]>(`${this.baseUrl}/vehicles/${userId}`);
  }

  // Upload Vehicle
  uploadVehicle(userId: number, vehicleNumber: string, owner: string, imageFile: File): Observable<Vehicle> {
    const formData = new FormData();
    formData.append('user_id', userId.toString());
    formData.append('number', vehicleNumber);
    formData.append('owner', owner);
    formData.append('image', imageFile);

    return this.http.post<Vehicle>(`${this.baseUrl}/upload/`, formData);
  }

  // Delete Vehicle
  deleteVehicle(vehicleId: number): Observable<any> {
    return this.http.delete(`${this.baseUrl}/vehicle/${vehicleId}`);
  }

  // Get Image URL
  getImageUrl(vehicleId: number): string {
    return `${this.baseUrl}/image/${vehicleId}`;
  }
}
